### PR TITLE
Redux/answerSlice

### PR DESCRIPTION
**Pull request by HossamOkasha**
Add Redux Answer Slice.

### CODE REVIEW

File: src/apis/answers.ts [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+// general types used here are defined in ./src/types.d.ts file
+type Normalized<T = Result> = {
+  entities: {
+    answers: Record<number, Answer>;
+    users: Record<string, User>;
+  };
+  result: T;
+};
+
+type fetchQAArgs = {qId: number, page: number};
```
Comment (ahmedhrayyan) [inline src/apis/answers.ts:15]: * types should be UpperCamelCased 
* also here this type should be singular as it only describe one arg
```suggestion
type FetchQAArg = {qId: number, page: number};
```

File: src/apis/answers.ts [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+async function fetchQuestionAnswers({qId, page}: fetchQAArgs) {
+  const { data } = await client.get(`/questions/${qId}/answers`, {
+    params: { page }
+  });
```
Comment (ahmedhrayyan) [inline src/apis/answers.ts:19]: Inject the page directly as you are already using string literals
```suggestion
  const { data } = await client.get(`/questions/${qId}/answers?page=${page}`)
```

File: src/apis/answers.ts [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  return normalize(data, { data: [aEntity] }) as Normalized;
+}
+
+async function show(id: number) {
+  const { data } = await client.get(`/answers/${id}`);
+  return normalize(data, { data: aEntity }) as Normalized<Omit<Result, "meta">>;
+}
+
+type storeArgs = {content: string, question_id: number};
```
Comment (ahmedhrayyan) [inline src/apis/answers.ts:28]: The same as above

File: src/redux/slices/answerSlice.ts [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+  createEntityAdapter,
+  createSlice,
+  isFulfilled,
+  isPending,
+  isRejected,
+} from "@reduxjs/toolkit";
+import aApi from "../../apis/answers";
+import { changeVote } from "../../utils/redux";
+
+type pageArg = { qId: number, page: number }
```
Comment (ahmedhrayyan) [inline src/redux/slices/answerSlice.ts:12]: unused type

File: src/redux/slices/answerSlice.ts [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+export const handleLoadAnswers = createAsyncThunk("a/all",
+  aApi.fetchQuestionAnswers, {
+  condition: ({qId, page}, { getState }) =>
+    !getState().questions.entities[qId]?.fetchedAnsPages?.includes(page),
+});
+
+export const handleShowAnswer = createAsyncThunk("a/show", aApi.show, {
+  condition: (id, { getState }) => !getState().answers.ids.includes(id),
```
Comment (ahmedhrayyan) [inline src/redux/slices/answerSlice.ts:20]: Good job here

File: src/redux/slices/answerSlice.ts [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+  (a: Answer) => aApi.remove(a.id)
+);
+
+type VoteArg = { answer: Answer; vote: Vote };
+export const handleVoteAnswer = createAsyncThunk(
+  "a/vote",
+  ({ answer, vote }: VoteArg) => aApi.vote(answer.id, vote)
+);
+
+type ansArg = { content: string, q_id: number };
```
Comment (ahmedhrayyan) [inline src/redux/slices/answerSlice.ts:34]: unused type

File: src/redux/slices/answerSlice.ts [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+  reducers: {
+    answerAdded: aAdapter.upsertOne,
+    answerRemoved: aAdapter.removeOne,
+  },
+  extraReducers: (builder) => {
+    builder
+      .addCase(handleLoadAnswers.fulfilled, (state, { payload }) => {
+        state.status = "succeeded";
+        state.total = payload.result.meta.total;
+        // state.fetchedPages.push(payload.result.meta.current_page);
```
Comment (ahmedhrayyan) [inline src/redux/slices/answerSlice.ts:56]: Remove this useless line

File: src/redux/slices/answerSlice.ts [hunk 8]
```diff
@@ -0,0 +1,109 @@
...
+  }),
+  reducers: {
+    answerAdded: aAdapter.upsertOne,
+    answerRemoved: aAdapter.removeOne,
+  },
+  extraReducers: (builder) => {
+    builder
+      .addCase(handleLoadAnswers.fulfilled, (state, { payload }) => {
+        state.status = "succeeded";
+        state.total = payload.result.meta.total;
```
Comment (ahmedhrayyan) [inline src/redux/slices/answerSlice.ts:55]: Remove total as it is meaningless in answers

File: src/redux/slices/answerSlice.ts [hunk 9]
```diff
@@ -0,0 +1,109 @@
...
+type ansArg = { content: string, q_id: number };
+export const handleAddAnswer = createAsyncThunk("a/add",  aApi.store);
+export const handleUpdateAnswer = createAsyncThunk("a/update", aApi.update);
+
+const aAdapter = createEntityAdapter<Answer>();
+
+const slice = createSlice({
+  name: "answers",
+  initialState: aAdapter.getInitialState({
+    total: 0,
```
Comment (ahmedhrayyan) [inline src/redux/slices/answerSlice.ts:43]: Remove total as it is meaningless in answers

File: src/types.d.ts [hunk 10]
```diff
@@ -26,6 +26,7 @@ type Question = {
 	upvotes: number;
 	user: number;
 	viewer_vote: null | boolean;
+	fetchedAnsPages?: number[]
```
Comment (ahmedhrayyan) [inline src/types.d.ts:29]: Make this fetchedAPages to keep consistency and add a comment to it identifying its purpose

**Comment by ahmedhrayyan** [comment]
Also please format all your code using `Prettier` before you push your changes

**Comment by HossamOkasha** [comment]
@ahmedhrayyan  Done, may you review it again?

**Pull request merged**